File: src/main/go/leetcode/lists/hot100/LeetCode23.go
package hot100

func mergeKLists(lists []*ListNode) *ListNode {
	if len(lists) == 0 {
		return nil
	}
	var mergeKLists func([]*ListNode, int, int) *ListNode
	mergeKLists = func(lists []*ListNode, left int, right int) *ListNode {
		if left >= right {
			return lists[left]
		}
		mid := (right - left) / 2
		list1 := mergeKLists(lists, left, mid)
		list2 := mergeKLists(lists, mid+1, right)
		return mergeList(list1, list2)
	}
	return mergeKLists(lists, 0, len(lists)-1)
}

File: src/main/java/leetcode/lists/topinterview/LeetCode23.java
package leetcode.lists.topinterview;

import common.*;

/**
 * @author wujingxinit@outlook.com
 * @date 2022/9/21 01:33
 */
public class LeetCode23 {

    public ListNode mergeKLists(ListNode[] lists) {
        if (lists == null || lists.length == 0) {
            return null;
        }
        return mergeKLists(lists, 0, lists.length - 1);
    }

    public ListNode mergeKLists(ListNode[] lists, int left, int right) {
        if (left >= right) {
            return lists[left];
        }
        int mid = (left + right) / 2;
        ListNode list1 = mergeKLists(lists, left, mid);
        ListNode list2 = mergeKLists(lists, mid + 1, right);
        return mergeTwoLists(list1, list2);
    }

    public ListNode mergeTwoLists(ListNode list1, ListNode list2) {
        if (list1 == null) {
            return list2;
        }
        if (list2 == null) {
            return list1;
        }

        ListNode dummy = new ListNode();
        ListNode cur = dummy;

        while (list1 != null && list2 != null) {
            if (list1.val < list2.val) {
                cur.next = list1;
                list1 = list1.next;
            } else {
                cur.next = list2;
                list2 = list2.next;
            }
            cur = cur.next;
        }
        if (list1 != null) {
            cur.next = list1;
        }
        if (list2 != null) {
            cur.next = list2;
        }
        return dummy.next;
    }
}
